### ISSUE TITLE
save/load model

### ISSUE THREAD

**Issue by fhowen** [dmlc/mxnet#479]
hi,if I want to save/load model during the trainning,which function should I use ? and can I set it to atuo-save every n epochs?
thanks

**Comment by winstywang** [comment]
You may need a callback at the end of each epoch like: https://github.com/dmlc/mxnet/blob/master/example/imagenet/inception-full.py#L101

**Comment by fhowen** [comment]
thank you,now I understand how to save the model as I like,but how about load model?
(1)if I want to load model to prediction,should I use mx.model.load?
(2)what about I load the model to continue trainning?

**Comment by tornadomeet** [comment]
(1), Yes
(2), refer to #379

**Comment by fhowen** [comment]
@tornadomeet thank you
I followed your advice and add this in the mx.model.FeedForward:arg_params=mx.nd.load("lenet_2GPU-0005.params"),and others in lenet.py stays unchanged,than I try it.
but it raise the error "find name "arg:fullyconnected1_bias" that is not in the arguments".
what's wrong in this?

**Comment by tornadomeet** [comment]
it seems that your *.params file and *-symbol.json file is inconsistent(the arg name : rg:fullyconnected1_bias).
list the args of symbol by list_arguments(), and list the dict of *.params to see the key, then check that

**Comment by fhowen** [comment]
from data import mnist_iterator
import mxnet as mx
import logging

## define lenet
# input
data = mx.symbol.Variable('data')
.
.
.
fc2 = mx.symbol.FullyConnected(data=tanh3, num_hidden=10)
# loss
lenet = mx.symbol.SoftmaxOutput(data=fc2)

## data
train, val = mnist_iterator(batch_size=300, input_shape=(1,28,28))

## train

logging.basicConfig(level=logging.DEBUG)
dev = [mx.gpu(i) for i in range(2)]
#dev = mx.gpu()

model = mx.model.FeedForward(
    ctx = dev, symbol = lenet, num_epoch = 5,
    learning_rate = 0.05, momentum = 0.9, wd = 0.00001,arg_params=mx.nd.load("lenet_2GPU-0005.params"))

#model=mx.model.load_checkpoint("lenet_2GPU",5)
model.fit(X=train, eval_data=val,
          batch_end_callback=mx.callback.Speedometer(300),
          epoch_end_callback=mx.callback.do_checkpoint("lenet_2GPU"))
#mx.model.save_checkpoint("lenet",1,lenet,lenet.arg_parans,lenet.aux_params)


this is lenet.py,I haven't use *-symbol.json,it can't be done liks this?

**Comment by fhowen** [comment]
@tqchen 
is there a bug? in excutor.py ,def copy_params_from()?

because if we use save_checkpoint() to save the model,the name of args will be like this "arg:fullyconnected1_bias"
but when we reload the model,the name in self.arg_dict,it doesn't have a prefix "arg:",
I think it is why I come to an error

**Comment by fhowen** [comment]
@tqchen
is there a bug? in excutor.py ,def copy_params_from()?

because if we use save_checkpoint() to save the model,the name of args will be like this "arg:fullyconnected1_bias"
but when we reload the model,the name in self.arg_dict,it doesn't have a prefix "arg:",
I think it is why I come to an error

**Comment by fhowen** [comment]
I write the forward like this:
model = mx.model.FeedForward(
        ctx = dev, symbol =mx.symbol.load('lenet_2GPU-symbol.json'), num_epoch = 10, begin_epoch=5,
        learning_rate = 0.05, momentum = 0.9, wd = 0.00001,arg_params=mx.nd.load('lenet_2GPU-0005.params'))

then I call fit.

**Comment by tornadomeet** [comment]
@fhowen  does it work now?

**Comment by fhowen** [comment]
@tornadomeet 
https://github.com/dmlc/mxnet/blob/master/python/mxnet/executor.py#L180
this is where error occurs.

**Comment by fhowen** [comment]
it doesn't work now..

**Comment by fhowen** [comment]
the name in self.arg_dict has no prefix,it comes from the symbol file 
but the name in arg_params has a prefix,it comes from arg_params file.

**Comment by fhowen** [comment]
https://github.com/dmlc/mxnet/blob/master/python/mxnet/model.py#L418
but if I remove the arg: here.
when I reload the model,it can only continue to train 1 epoch ,then an error of save_dict.update occurs.

**Comment by tornadomeet** [comment]
@fhowen  i just run a test, but no that bug here, so can u paste your code?

**Comment by fhowen** [comment]
ok.

**Comment by fhowen** [comment]
# pylint: skip-file
from data import mnist_iterator
import mxnet as mx
import logging

## define lenet
# input
data = mx.symbol.Variable('data')
"""
# first conv
conv1 = mx.symbol.Convolution(data=data, kernel=(5,5), num_filter=20)
tanh1 = mx.symbol.Activation(data=conv1, act_type="tanh")
pool1 = mx.symbol.Pooling(data=tanh1, pool_type="max",
                          kernel=(2,2), stride=(2,2))
# second conv
conv2 = mx.symbol.Convolution(data=pool1, kernel=(5,5), num_filter=50)
tanh2 = mx.symbol.Activation(data=conv2, act_type="tanh")
pool2 = mx.symbol.Pooling(data=tanh2, pool_type="max",
                          kernel=(2,2), stride=(2,2))
# first fullc
flatten = mx.symbol.Flatten(data=pool2)
fc1 = mx.symbol.FullyConnected(data=flatten, num_hidden=500)
tanh3 = mx.symbol.Activation(data=fc1, act_type="tanh")
# second fullc
fc2 = mx.symbol.FullyConnected(data=tanh3, num_hidden=10)
# loss
lenet = mx.symbol.SoftmaxOutput(data=fc2)
"""
## data
train, val = mnist_iterator(batch_size=300, input_shape=(1,28,28))

## train

logging.basicConfig(level=logging.DEBUG)
dev = [mx.gpu(i) for i in range(2)]
#dev = mx.gpu()
prefix="lenet_2GPU"

model = mx.model.FeedForward(
        ctx = dev, symbol =mx.symbol.load('lenet_2GPU-symbol.json'), num_epoch = 10, begin_epoch=5,
        learning_rate = 0.05, momentum = 0.9, wd = 0.00001,arg_params=mx.nd.load('lenet_2GPU-0005.params'))

#model = mx.model.FeedForward(
#    ctx = dev, symbol =lenet, num_epoch = 5,
#    learning_rate = 0.05, momentum = 0.9, wd = 0.00001)
#model=mx.model.load_checkpoint("lenet_2GPU",5)
model.fit(X=train, eval_data=val,
          batch_end_callback=mx.callback.Speedometer(300),
          epoch_end_callback=mx.callback.do_checkpoint(prefix))
#mx.model.save_checkpoint("lenet",1,lenet,lenet.arg_parans,lenet.aux_params)
#model.save(prefix,10)

**Comment by fhowen** [comment]
@tornadomeet sorry for the code is messy...I haven't used github before

**Comment by tornadomeet** [comment]
@fhowen , hmm, this seems a bug ;
the reason is that we cannot use "arg_params=mx.nd.load('lenet_2GPU-0005.params')" directly, because mx.nd.load() is return not just arg_params but also aux_params, so we should seperate the arg_params from the return, and then feed into mx.model.FeedForward(...)
acturally, the mx.nd.load() interface is a bit puzzling for users, maybe @tqchen , can add seperate interface, like mx.nd.load_arg(), mx.nd.load_aus()

**Comment by fhowen** [comment]
@tornadomeet thank you.
I hope there is a example about continue to train the model...

**Comment by tqchen** [comment]
You can load it in a model structure, and feed the content to another
```python
loaded = mx.model.load('xxx', 10)

continue_model = mx.model.FeedForward(symbol = loaded.symbol, arg_params = loaded.arg_params, aux_params = loaded.aux_params, other parameters)

continue_model.fit()
```

**Comment by tqchen** [comment]
closing for now as answered in last comment, please feel free to reopen

**Comment by kvazipushka** [comment]
Hello, @tqchen . Could you please provide information or example, how can I continue training network in R? I was able to find there only mx.model.FeedForward.create() without 'arg_params' as a parametrs. Thanks.

**Comment by maxkazmsft** [comment]
arg.params and aux.params have been added - it's possible to continue training the model, but you have to re-specify the other training parameters, like learning.rate, X, etc.

Is there a way to _just_ reload trained model for scoring? mx.model.FeedForward.create will initiate training anyway - I just want to load.
